Add serverless support

Bouncheck opened the pull request:
Contains most of the recent improvements in regards to serverless Scylla support. 
The driver should be able to correctly read .yaml configuration file and establish connection
with a cluster through the SNI proxy.

Bouncheck committed:
Add classes modelling Scylla Cloud configuration

Contains methods that allow for initialization with .yaml configuration file.
Adds Jackson library as necessary dependency.
Adds Bouncy Castle as necessary dependency due to keys from configuration file
possibly being in `PKCS#1` format.

Bouncheck committed:
Add Scylla Cloud .yaml parsing test

This test class parses two hard coded configuration files
and tries to instantiate all necessary classes with provided info
but without establishing trial connection.

Bouncheck committed:
Add testing support and basic tests for Scylla Cloud

Extends `@CCMConfig` annotation with new option to start SNI proxy.
Adds custom `clusterProvider` for tests with proxy enabled.
It uses special `MockSniEndPointFactory` in order to resolve hosts
correctly on local network.

The tests establish connection with single and multi node
SNI proxy enabled ccm clusters using ccm generated .yaml for configuration.
Simple queries like keyspace and table creation are executed and verified.

Adding nodes to already running proxy enabled ccm clusters is not working yet.
However as a side effect of the changes some `long` tests that do add nodes
to clusters without proxy have been fixed.

Bouncheck committed:
Add `@ScyllaOnly` annotation

Needed for Scylla Cloud tests which won't work with cassandra ccm clusters.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ConfigurationBundle.java:
```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
```
It's not a full copyright comment.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ConfigurationBundle.java:
```diff
@@ -0,0 +1,120 @@
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509TrustManager;
+
+public class ConfigurationBundle {
```
```suggestion
class ConfigurationBundle {
```

Let's make it private (if possible). The reason is that we can more easily modify that in the future without worrying about backwards compatiblity or someone manually using `ConfigurationBundle` which is an object for internal use only.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ConfigurationBundle.java:
```diff
@@ -0,0 +1,120 @@
+      throws IOException, GeneralSecurityException {
+    KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+    kmf.init(ks, "cassandra".toCharArray());
+    // Arrays.fill(keyStorePassword, (char) 0);
```
This should be deleted.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ConfigurationBundle.java:
```diff
@@ -0,0 +1,120 @@
+    return tmf;
+  }
+
+  protected SSLOptions getSSLOptions() {
+    try {
+      return SniSSLOptions.builder().withSSLContext(getSSLContext()).build();
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  protected SSLOptions getInsecureSSLOptions() {
+    try {
+      return SniSSLOptions.builder().withSSLContext(getInsecureSSLContext()).build();
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
```
Why those `try {} catch() {}`?

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/Parameters.java:
```diff
@@ -0,0 +1,20 @@
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class Parameters {
```
Similarly to my previous comment about removing `public`.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/Parameters.java:
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (C) 2022 ScyllaDB
+ */
```
This is not a full copyright comment.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudAuthInfo.java:
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2022 ScyllaDB
+ */
```
This is not a full copyright comment.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudAuthInfo.java:
```diff
@@ -0,0 +1,91 @@
+    }
+  }
+
+  byte[] getClientCertificateData() {
```
Missing `public` modifier (there and in 3 methods below).

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudConnectionConfig.java:
```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (C) 2022 ScyllaDB
+ */
```
Not a full copyright comment.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudConnectionConfig.java:
```diff
@@ -0,0 +1,202 @@
+      throw new IllegalArgumentException(
+          "Please provide datacenters (datacenters:) in the configuration yaml.");
+    }
+    for (ScyllaCloudDatacenter x : datacenters.values()) {
```
Small nit: rename `x` to `datacenter`.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudConnectionConfig.java:
```diff
@@ -0,0 +1,202 @@
+      throw new IllegalArgumentException(
+          "Please provide any authentication config (authInfos:) in the configuration yaml.");
+    }
+    for (ScyllaCloudAuthInfo x : authInfos.values()) {
```
Small nit: rename `x` to `authInfo`.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudConnectionConfig.java:
```diff
@@ -0,0 +1,202 @@
+    this.validate();
+    KeyStore identity = KeyStore.getInstance(KeyStore.getDefaultType());
+    KeyStore trustStore = KeyStore.getInstance(KeyStore.getDefaultType());
+    identity.load(null, "cassandra".toCharArray());
+    trustStore.load(null, "cassandra".toCharArray());
```
If possible, get rid of hardcoded `cassandra`.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudConnectionConfig.java:
```diff
@@ -0,0 +1,202 @@
+    identity.load(null, "cassandra".toCharArray());
+    trustStore.load(null, "cassandra".toCharArray());
+
+    for (Map.Entry<String, ScyllaCloudDatacenter> x : datacenters.entrySet()) {
```
Small nit: rename `x` to `datacenter`.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudContext.java:
```diff
@@ -0,0 +1,26 @@
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class ScyllaCloudContext {
```
Similar problem with `public` as I commented before.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudDatacenter.java:
```diff
@@ -0,0 +1,83 @@
+    }
+  }
+
+  String getCertificateAuthorityPath() {
```
Missing `public` modifier here and in method below.

avelanarius commented on driver-core/src/test/java/com/datastax/driver/core/CloudConfigYamlParsingTest.java:
```diff
@@ -0,0 +1,40 @@
+
+public class CloudConfigYamlParsingTest {
+
+  private static String CONFIG_PATH = "/scylla_cloud/testConf.yaml";
```
If `CONFIG_PATH` is overwritten in each test, maybe it should be removed.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -1393,92 +1385,52 @@ public Builder withNoCompact() {
       return this;
     }
 
-    /**
```
Removing of `failIfCloud` and `withCloudSecureConnectBundle` should be probably a part of `
Remove Datastax cloud implementation methods` commit.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:
```diff
@@ -298,6 +298,9 @@ ListenableFuture<Void> initAsyncWithConnection(Connection reusedConnection) {
     connectionsPerShard = coreSize / shardsCount + (coreSize % shardsCount > 0 ? 1 : 0);
     maxConnectionsPerShard =
         maxConnections / shardsCount + (maxConnections % shardsCount > 0 ? 1 : 0);
+    System.out.printf(
+        "connectionsPerShard %d, maxConnectionPerShard %d, shardCount %d\n",
+        connectionsPerShard, maxConnectionsPerShard, shardsCount);
```
This is a debug print, it should be removed.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/SniSSLOptions.java:
```diff
@@ -92,7 +97,10 @@ protected SSLEngine newSSLEngine(
     // To avoid that, we create a unique "fake" port for every node. We still get session reuse for
     // a given node, but not across nodes. This is safe because the advisory port is only used for
     // session caching.
-    SSLEngine engine = context.createSSLEngine(address.getHostName(), getFakePort(sniServerName));
+
+    // SSLEngine engine = context.createSSLEngine(address.getHostName(),
+    // getFakePort(sniServerName));
+    SSLEngine engine = context.createSSLEngine(sniServerName, getFakePort(sniServerName));
```
Wasn't this something done only for tests and now can be removed?

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -192,8 +191,8 @@ private Cluster(
       Collection<Host.StateListener> listeners) {
-    System.out.println("===== Using optimized driver!!! =====");
-    logger.info("===== Using optimized driver!!! =====");
+    System.out.println("===== Using Cloud optimized driver!!! =====");
```
(not commenting on this specific line) Please add some description to `Enable connectivity with Scylla Cloud clusters` commit.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/CloudConfig.java:
```diff
@@ -1,70 +0,0 @@
-/*
```
You could mention https://github.com/scylladb/java-driver/commit/c2d725f0ddd6fba2412567f40bd92de98adc9124 in description of "Remove Datastax cloud implementation methods" commit - it'd give some additional context about the removal (the documentation of the feature was already removed before).

avelanarius commented on clirr-ignores.xml:
```diff
@@ -16,6 +16,27 @@
     Modified by ScyllaDB
 -->
 <differences>
+    <difference>
```
This should be squashed with "Remove Datastax cloud implementation methods" commit.

avelanarius commented on driver-core/src/test/java/com/datastax/driver/core/CCMConfig.java:
```diff
@@ -210,6 +215,9 @@ final class Undefined {}
    */
   boolean[] createKeyspace() default {};
 
+  /* nocheckin document */
```
My leftover comment...

avelanarius commented on driver-core/src/test/java/com/datastax/driver/core/MockSniEndPointFactory.java:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Copyright (C) 2022 ScyllaDB
+ *
+ * Modified by ScyllaDB
+ */
```
This shouldn't have `Copyright DataStax, Inc.` - it was fully done by us.

Bouncheck committed:
Enable connectivity with Scylla Cloud clusters

Adds `ScyllaCloudSniEndPointFactory` and several `Cluster.Builder`
methods for initializing with `ScyllaCloudConnectionConfig`.

Bouncheck committed:
Remove Datastax's cloud implementation methods

Those won't be maintained in this repository
and will only create confusion.

Adds clirr ignores for removed test methods.

Documentation part was already deleted by c2d725f0ddd6fba2412567f40bd92de98adc9124

zimnx commented:
Was it tested on serverless env? Our dev env is down for quite a while.

zimnx commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -192,8 +191,8 @@ private Cluster(
       Collection<Host.StateListener> listeners) {
-    System.out.println("===== Using optimized driver!!! =====");
-    logger.info("===== Using optimized driver!!! =====");
+    System.out.println("===== Using Cloud optimized driver!!! =====");
```
would be better to distinguish between direct connection and cloud connection and print out "Cloud" only when latter.

fruch replied:
I would be nice if this string isn't changed at all.

Since there is code in SCT using this line to tell if it's shared aware driver or not

I don't see any value to knowing if it's cloud related or not (i.e. whom run it with a bundle would know with any log)

zimnx commented on driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java:
```diff
@@ -668,6 +670,11 @@ private static void updateInfo(
     host.setHostId(row.getUUID("host_id"));
     host.setSchemaVersion(row.getUUID("schema_version"));
+
+    EndPoint endPoint = endPointForPeerHost(row, host.getEndPoint(), cluster);
```
This funciton is essentialy a one liner for Scylla (it handles some specific behavior of datastax), it could be inlined here.

zimnx commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudDatacenter.java:
```diff
@@ -0,0 +1,83 @@
+                + ").");
+      }
+    }
+    if (server == null) {
+      throw new IllegalArgumentException("server property is required in datacenter description.");
+    }
```
validate wheter it's a URL with port

zimnx commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudDatacenter.java:
```diff
@@ -0,0 +1,83 @@
+    if (server == null) {
+      throw new IllegalArgumentException("server property is required in datacenter description.");
+    }
+    if (nodeDomain == null) {
+      throw new IllegalArgumentException(
+          "nodeDomain property is required in datacenter description.");
+    }
```
validate whether it's a DNS subdomain

zimnx commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudDatacenter.java:
```diff
@@ -0,0 +1,83 @@
+      @JsonProperty(value = "certificateAuthorityPath") String certificateAuthorityPath,
+      @JsonProperty(value = "certificateAuthorityData") byte[] certificateAuthorityData,
+      @JsonProperty(value = "server") String server,
+      @JsonProperty(value = "TLSServerName") String TLSServerName,
```
it's kept as `tlsServerName` field in yaml,

zimnx reviewed (commented):
I recall Java driver puts initial contact point into host pool. This leads to bad performance, because every request to it land on different node and breaks node awareness. 
I don't see any changes related to it in this PR.

zimnx commented on driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java:
```diff
@@ -799,6 +806,8 @@ private void refreshNodeListAndTokenMap(
     } else {
       updateInfo(controlHost, localRow, cluster, isInitialConnection);
+      connection.endPoint = controlHost.getEndPoint();
+
```
This function should also refresh metada from local table, not just peers.

fruch commented:
@Bouncheck 
you still need the pkcs8 key ? (notice I've move it in)
https://github.com/scylladb/scylla-ccm/pull/399

Bouncheck commented:
> @Bouncheck you still need the pkcs8 key ? (notice I've move it in) [scylladb/scylla-ccm#399](https://github.com/scylladb/scylla-ccm/pull/399)

Not anymore. I've ended up adding a dependency for handling that anyway. Currently code here is expecting standard pkcs1 key but i intend to update it so that driver will handle both versions. I will push it together with other fixes.
Sorry for the trouble.

fruch commented:
> I've ended up adding a dependency for handling that anyway. Currently code here is expecting standard pkcs1 key but i intend to update it so that driver will handle both versions. I will push it together with other fixes.
> Sorry for the trouble.

just want to make sure I didn't broke your testing code.
and glad to see the ccm code is working as expected

fruch reviewed (approved):
LGTM

Bouncheck commented:
I've force pushed v2.
Things addressed:
System.out.println with cloud driver message was reverted,
Improved URL validation,
Improved nodeDomain validation,
`@JsonProperty tlsServerName`,
 inlined `endPointForPeerHost` call
 
 Regarding `refreshNodeListAndTokenMap(` I see it does use a row from system.local to update stuff. I do not know if it's missing something I'm overlooking. So far I haven't experienced issues and all hosts in tests do end up getting discovered with correct host ids.
 Regarding putting initial contact point in the pool, the current code updates connection endpoint after looking at system.local table of node it gets redirected to. I've also added additional assert to the test that checks every connected Host's EndPoint if it contains "any." substring. It seems both v1 and v2 successfully avoids adding "any" endpoint to discovered hosts.

Changes were checked mostly using CCM tests and some manual checks on dev cluster recently.

zimnx commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudDatacenter.java:
```diff
@@ -0,0 +1,113 @@
+      throw new IllegalArgumentException(
+          "nodeDomain property is required in datacenter description.");
+    } else {
+      if (nodeDomain.contains(" "))
```
According to RFC 1123 a lowercase subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character. These condintions aren't checking it

Bouncheck replied:
It seems i mixed up 'domain names' with 'hostnames' which have different rules. Will fix that.

zimnx commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudDatacenter.java:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  public String getTLSServerName() {
+    return TLSServerName;
```
this field start with upper case letter contrary to how other fields are called. I understand it's a abbreviation, but for SNI this uppercase isn't used. Would be good to be consistent.

zimnx commented on driver-core/src/main/java/com/datastax/driver/core/ScyllaCloudSniEndPointFactory.java:
```diff
@@ -0,0 +1,45 @@
+  public void init(Cluster cluster) {}
+
+  @Override
+  public EndPoint create(Row peersRow) {
```
in theory this can be a localRow aswell

Bouncheck commented:
Force pushed v3:
`TLSServerName` -> `tlsServerName`,
`peersRow` -> `row`,
`nodeDomain` is now validated more strictly in line with RFC-1123

zimnx reviewed (approved):
LGTM, thanks
